File: 3nplus1/solution.js

var fs = require("fs");
var input = fs.readFileSync(0, "utf-8").trim().split("\n");
var memo = new Map();
memo.set(1, 1);
function cycleLength(n) {
    if (memo.has(n))
        return memo.get(n);
    var length;
    if (n % 2 === 0)
        length = 1 + cycleLength(n / 2);
    else
        length = 1 + cycleLength(3 * n + 1);
    memo.set(n, length);
    return length;
}
for (var _i = 0, input_1 = input; _i < input_1.length; _i++) {
    var line = input_1[_i];
    if (!line.trim())
        continue;
    var _a = line.split(" "), iStr = _a[0], jStr = _a[1];
    var i = parseInt(iStr);
    var j = parseInt(jStr);
    var start = Math.min(i, j);
    var end = Math.max(i, j);
    var maxCycle = 0;
    for (var k = start; k <= end; k++) {
        maxCycle = Math.max(maxCycle, cycleLength(k));
    }
    console.log("".concat(i, " ").concat(j, " ").concat(maxCycle));
}

File: 3nplus1/solution.c
#include <stdio.h>

#define MAX 1000000

long memo[MAX] = {0};

long cycleLength(long n ) {
    if(n==1)
        return 1;

    if(n<MAX && memo[n] != 0)
        return memo[n];

    long length;
    if(n%2==0)
        length = 1+cycleLength(n/2);

    else
        length = 1+cycleLength(3*n+1);

    if(n<MAX) 
        memo[n] = length;

    return length;
}

int main() {
    long i,j;

    while(scanf("%ld %ld",&i,&j)==2) {
        long start = i<j ? i:j;
        long end = i>j ? i:j;

        long maxCycle = 0;

        for(long k=start;k<=end;k++) {
            long length = cycleLength(k);
            if(length>maxCycle)
                maxCycle = length;
        }

        printf("%ld %ld %ld",i,j,maxCycle);
    }
    return 0;
}